core/src/main/java/org/apache/iceberg/BaseFilesTable.java

```diff
@@ -223,34 +226,37 @@ ManifestFile manifest() {
   static class ContentFileStructWithMetrics implements StructLike {
     private final StructLike fileAsStruct;
     private final MetricsUtil.ReadableMetricsStruct readableMetrics;
-    private final int expectedSize;
+    private final int columnCount;
+    private final int metricsPosition;
 
     ContentFileStructWithMetrics(
-        int expectedSize,
+        int columnCount,
+        int metricsPosition,
         StructLike fileAsStruct,
         MetricsUtil.ReadableMetricsStruct readableMetrics) {
       this.fileAsStruct = fileAsStruct;
       this.readableMetrics = readableMetrics;
-      this.expectedSize = expectedSize;
+      this.columnCount = columnCount;
+      this.metricsPosition = metricsPosition;
     }
 
     @Override
     public int size() {
-      return expectedSize;
+      return columnCount;
     }
 
     @Override
     public <T> T get(int pos, Class<T> javaClass) {
-      int lastExpectedIndex = expectedSize - 1;
-      if (pos < lastExpectedIndex) {
+      if (pos < metricsPosition) {
         return fileAsStruct.get(pos, javaClass);
-      } else if (pos == lastExpectedIndex) {
+      } else if (pos == metricsPosition) {
         return javaClass.cast(readableMetrics);
       } else {
-        throw new IllegalArgumentException(
-            String.format(
-                "Illegal position access for ContentFileStructWithMetrics: %d, max allowed is %d",
-                pos, lastExpectedIndex));
+        // columnCount = fileAsStruct column count + the readable metrics field.
+        // When pos is greater than metricsPosition, the actual position of the field in
+        // fileAsStruct should be
+        // subtracted by 1.
```

Comment:
nit: merge this line with the line above

Reply:
done